tests/Database/DatabaseEloquentBelongsToManyTest.php

```diff
@@ -609,53 +608,162 @@ public function testTouchIfTouching()
         $relation->touchIfTouching();
     }
 
-    public function testSyncMethodConvertsCollectionToArrayOfKeys()
+    public function testSyncMethodAcceptsCollectionOfModels()
     {
-        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching', 'formatRecordsList'])->setConstructorArgs($this->getRelationArguments())->getMock();
-        $query = m::mock('stdClass');
-        $query->shouldReceive('from')->once()->with('user_role')->andReturn($query);
-        $query->shouldReceive('where')->once()->with('user_id', 1)->andReturn($query);
-        $relation->getQuery()->shouldReceive('getQuery')->andReturn($mockQueryBuilder = m::mock('StdClass'));
-        $mockQueryBuilder->shouldReceive('newQuery')->once()->andReturn($query);
-        $query->shouldReceive('pluck')->once()->with('role_id')->andReturn(new BaseCollection([1, 2, 3]));
+        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching'])->setConstructorArgs($this->getRelationArguments())->getMock();
+        $query = $this->getNewQuery($relation);
+        $query->shouldReceive('pluck')->once()->with('role_id')->andReturn(new BaseCollection());
 
         $collection = new Collection([
             m::mock(['getKey' => 1]),
             m::mock(['getKey' => 2]),
-            m::mock(['getKey' => 3]),
         ]);
-        $relation->expects($this->once())->method('formatRecordsList')->with([1, 2, 3])->will($this->returnValue([1 => [], 2 => [], 3 => []]));
-        $relation->sync($collection);
+        $this->assertEquals(['attached' => [1, 2], 'detached' => [], 'updated' => []], $relation->sync($collection));
     }
 
     public function testWherePivotParamsUsedForNewQueries()
     {
-        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching', 'formatRecordsList'])->setConstructorArgs($this->getRelationArguments())->getMock();
+        $relation = $this->getMockBuilder('Illuminate\Database\Eloquent\Relations\BelongsToMany')->setMethods(['attach', 'detach', 'touchIfTouching'])->setConstructorArgs($this->getRelationArguments())->getMock();
+        $relation->getQuery()->shouldReceive('where')->once()->with('user_role.foo', '=', 'bar', 'and')->andReturnSelf();
+        $currentQuery = $this->getNewQuery($relation);
+        $currentQuery->shouldReceive('pluck')->andReturn(new BaseCollection([1, 2, 3]));
 
-        // we expect to call $relation->wherePivot()
-        $relation->getQuery()->shouldReceive('where')->once()->andReturn($relation);
+        // This is our test! The wherePivot() params also need to be called
```

Comment:
We should ask @GrahamCampbell :)
I guess it refers to the `wherePivot` line below. There are bunch of tests in this class that are mocking CUT and require refactoring, but I didn't want to do that, so PR doesn't get even bigger and harder to review.